juju/main.py

```diff
@@ -0,0 +1,149 @@
+def _handle_event_link(charm_dir, event_name):
+    """Create a symlink for a particular event.
+
+    charm_dir -- A root directory of the charm
+    event_name -- A name of the event for which to create a symlink.
+    """
+    event_hook_path = charm_dir / 'hooks' / event_name.replace('_', '-')
+    create_link = True
+    # Remove incorrect symlinks or files.
+    if event_hook_path.exists():
+        # Non-symlink entries and the ones not pointing to "install"
+        # need to be removed.
+        if not event_hook_path.is_symlink():
+            debugf(f'Hook entry at {event_hook_path} is not a symlink:'
+                   ' attempting to remove it.')
+            # May raise IsADirectoryError, e.g. in case it is a directory which
+            # is unexpected and left to the developer or operator to handle.
+            event_hook_path.unlink()
+        elif os.readlink(event_hook_path) != 'install':
+            debugf(f'Removing entry {event_hook_path} as it does not point'
+                   ' to "install"')
+            event_hook_path.unlink()
+        else:
+            create_link = False
+
+    if create_link:
+        debugf(f'Creating a new relative symlink at {event_hook_path}'
+               f' to pointing to "install" located in {charm_dir}/hooks')
+        event_hook_path.symlink_to('install')
+
+
+def _setup_hooks(charm_dir, charm):
+    """Set up hooks for supported events.
+
+    Whether a charm can handle an event or not can be determined by
+    introspecting which events are bound to it.
+
+    Hooks are created as symlinks to "install" which may be either
+    a copy of main.py (which calls the main function) or a symlink
+    to it. Note that it is important to avoid creating a recursive
+    symlink which is why the install event itself is skipped.
+
+    charm_dir -- A root directory of the charm.
+    charm -- An instance of the Charm class.
+    """
+    for event_name in charm.on.events().keys():
+        if event_name != 'install':
+            _handle_event_link(charm_dir, event_name)
+
+
+def _emit_charm_event(charm, event_name):
+    """Emits a charm event based on a Juju event name.
+
+    charm -- A charm instance to emit an event from.
+    event_name -- A Juju event name to emit on a charm.
+    """
+    formatted_event_name = event_name.replace('-', '_')
+    event_to_emit = None
+    try:
+        event_to_emit = getattr(charm.on, formatted_event_name)
+    except AttributeError:
+        debugf(f"event {formatted_event_name} not defined for {charm}")
+
+    # If the event is not supported by the charm implementation, do
+    # not error out or try to emit it. This is to support rollbacks.
+    if event_to_emit is not None:
+        debugf(f'Emitting Juju event {event_name}')
+        event_to_emit.emit()
+
+
+def main():
+    """Setup the charm and dispatch the observed event.
+
+    The event name is based on the way this executable was called (argv[0]).
+    """
+    import charm as charm_module
```

Comment:
If somebody ships an abstract charm as a module I guess they can use a different type name for it for others to create their `class Charm(AbstractCharm)`. Even then it can be imported as something else to be used for inheritance.

If we have multiple tests though that use different modules imported as `charm` for the module name then we will have to unload this module on `tearDown`.